contracts/defi/provider/UniswapProvider.sol

```diff
@@ -0,0 +1,155 @@
+ */
+contract UniswapProvider is Invest {
+
+    address constant internal ETH_TOKEN_ADDRESS = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+
+    using SafeMath for uint256;
+
+    /* ********************************** Implementation of Invest ************************************* */
+
+    /**
+     * @dev Invest tokens for a given period.
+     * @param _wallet The target wallet.
+     * @param _token The token address.
+     * @param _amount The amount of tokens to invest.
+     * @param _period The period over which the tokens may be locked in the investment (optional).
+     * @param _oracles (optional) The address of one or more oracles contracts that may be used by the provider to query information on-chain.
+     * @return The amount of tokens that have been invested. 
+     */
+    function addInvestment(
+        BaseWallet _wallet, 
+        address _token, 
+        uint256 _amount, 
+        uint256 _period, 
+        address[] calldata _oracles
+    ) 
+        external 
+        returns (uint256 _invested)
+    {
+        require(_oracles.length == 1, "Uniswap: invalid oracles length");
+        _invested = addLiquidity(_wallet, _oracles[0], _token, _amount);
+    }
+
+    /**
+     * @dev Removes a fraction of the tokens from an investment.
+     * @param _wallet The target wallet.s
+     * @param _token The array of token address.
+     * @param _fraction The fraction of invested tokens to exit in per 10000. 
+     * @param _oracles (optional) The address of one or more oracles contracts that may be used by the provider to query information on-chain.
+     */
+    function removeInvestment(
+        BaseWallet _wallet, 
+        address _token, 
+        uint256 _fraction, 
+        address[] calldata _oracles
+    ) 
+        external 
+    {
+        require(_fraction <= 10000, "Uniswap: _fraction must be expressed in 1 per 10000");
+        require(_oracles.length == 1, "Uniswap: invalid oracles length");
+        removeLiquidity(_wallet, _oracles[0], _token, _fraction);
+    }
+
+    /**
+     * @dev Get the amount of investment in a given token.
+     * @param _wallet The target wallet.
+     * @param _token The token address.
+     * @param _oracles (optional) The address of one or more oracles contracts that may be used by the provider to query information on-chain.
+     * @return The value in tokens of the investment (including interests) and the time at which the investment can be removed.
+     */
+    function getInvestment(
+        BaseWallet _wallet, 
+        address _token, 
+        address[] calldata _oracles
+    ) 
+        external 
+        view
+        returns (uint256 _tokenValue, uint256 _periodEnd) 
+    {
+        address tokenPool = UniswapFactory(_oracles[0]).getExchange(_token);
+        uint256 tokenPoolSize = ERC20(_token).balanceOf(tokenPool);
+        uint shares = ERC20(tokenPool).balanceOf(address(_wallet));
+        _tokenValue = shares.mul(tokenPoolSize).mul(2);
```

Comment:
Shouldn't we divide this by `ERC20(tokenPool).totalSupply()` to get the total invested value in token?

Reply:
Agreed and done.